test/spec/modules/uid2IdSystem_spec.js

```diff
@@ -342,4 +353,165 @@ describe(`UID2 module`, function () {
+            const bid = await runAuction();
+            expectNoIdentity(bid);
+            expectGlobalToHaveNoUid2();
+            expectModuleStorageEmptyOrMissing();
+          })
+
+          it('it should ignores token provided in server-set cookie', async function() {
+            cookieHelpers.setPublisherCookie(publisherCookieName, initialToken);
+            scenario.setConfig({ ...newServerCookieConfigParams, auctionDelay: 0, syncDelay: 1 })
+            const bid = await runAuction();
+            expectNoIdentity(bid);
+            expectGlobalToHaveNoUid2();
+            expectModuleStorageEmptyOrMissing();
+          })
+
+          describe('When the token generated in time', function() {
+            testApiSuccessAndFailure(async function(apiSucceeds) {
+              scenario.setConfig();
+              apiHelpers.respondAfterDelay(auctionDelayMs / 10, server);
+              const bid = await runAuction();
+
+              if (apiSucceeds) expectToken(bid, clientSideGeneratedToken);
+              else expectNoIdentity(bid);
+            }, cstgApiUrl, 'it should be used in the auction', 'the auction should have no uid2', false, clientSideGeneratedToken);
+
+            testApiSuccessAndFailure(async function(apiSucceeds) {
+              scenario.setConfig();
+              apiHelpers.respondAfterDelay(auctionDelayMs / 10, server);
+
+              await runAuction();
+              if (apiSucceeds) {
+                expectModuleStorageToContain(undefined, clientSideGeneratedToken, scenario.identity);
+              } else {
+                expectModuleStorageEmptyOrMissing();
+              }
+            }, cstgApiUrl, 'the generated token should be stored in the module storage', 'the module storage should not be set', false, clientSideGeneratedToken);
+          });
+        });
+      });
+    });
+    describe(`when the response doesn't arrive before the auction timer`, function() {
+      testApiSuccessAndFailure(async function() {
+        config.setConfig(makePrebidConfig({ ...cstgConfigParams, email: 'test@test.com' }));
+        const bid = await runAuction();
+        expectNoIdentity(bid);
+      }, cstgApiUrl, 'it should run the auction', undefined, false, clientSideGeneratedToken);
+
+      testApiSuccessAndFailure(async function(apiSucceeds) {
+        config.setConfig(makePrebidConfig({ ...cstgConfigParams, email: 'test@test.com' }));
+        const promise = apiHelpers.respondAfterDelay(auctionDelayMs * 2, server);
+
+        const bid = await runAuction();
+        expectNoIdentity(bid);
+        expectGlobalToHaveNoUid2();
+        await promise;
+        if (apiSucceeds) expectGlobalToHaveToken(clientSideGeneratedToken);
+        else expectGlobalToHaveNoUid2();
+      }, cstgApiUrl, 'it should update the userId after the auction', 'there should be no global identity', false, clientSideGeneratedToken);
+    })
+
+    describe('when there is a token in the module cookie', function() {
```

Comment:
does this cover if there's already uid2token via non-CSTG way, namely client-refresh mode and server-only mode mentioned [here](https://github.com/prebid/Prebid.js/blob/master/modules/uid2IdSystem.md)? 

Or in another word, do we have any test to verify the expected behaviour when pubs provides a mix of non-cstg and cstg param interchangeably? I'm not expecting a pub to do so intentionally but just wanna see if our code handles it?

Shall we have a separate set of unit tests for such combos:

1. If there's cstg token previously in storage and now pubs tries to pass us non cstg-derived UID2 token. We should always use the new non cstg-derived token? (as we can't compare hash?)
2. If there's non-cstg derived token and now the pub passes us cstg param (i think we can verify if the email/phone hash matches and if not then call cstg endpoint?)

Reply:
For pubs providing a mix of params, tests covers the following scenarios:
1. CSTG params are in config, non cstg-derived token stored, call API to generate new cstg-derived token (your case 2)
2. Both CSTG params and client-refresh or server-only params are given, ignores the non-cstg params
3. Given client-refresh or server-only params with no CSTG params, but cstg-derived token stored. As we compared the stored token is not equals to the pubs provided token, we will use the non cstg-derived token they provide (which is case 1)

I wonder if this covers all combos?

Reply:
Those scenario covers well and I think the only scenario remaining is if no params are given, but there are both cstg-derived and non-cstg-derived tokens are stored. Is this possible? If so, what should we do? Just pick whichever that hasn't expired and then whichever one can be refreshed?

Reply:
There will be only one kind of token stored. OriginalToken is either the token generated by cstgAPI or pubs provided, and latestToken is the token refreshed by OriginalToken.

I look into unit tests, we don't have anything that covers this scenario, I will definitely add that. Based on the code, if no token is provided and CSTG params is not given, we will return the stored token if it is still valid

Reply:
Thanks - so i think there are 2 new tests to be written:

1. If there's a cstg-derived token stored, and no new params given, then reuse the previously-stored cstg-derived token
2.  If there's a non-cstg-derived token stored, and no new params given, then reuse the previously-stored non-cstg-derived token

Reply:
When no params are given, we don't compare the stored value with params, so we don't really check the originalIdentity field anymore, do we need separate tests for cstg-dervied token and non-cstg-derived token?

The three tests I added are:

- when no params given and stored token is valid, use stored token
- when no params given and stored token is invalid, should not include uid2 in auction
- when no params and no stored token, should not include uid2 in auction

Reply:
_do we need separate tests for cstg-dervied token and non-cstg-derived token?_

For completeness it would be great if we could so we know for sure it all works as we expected. If too hard then maybe ignore it.

Reply:
I see, I've added test for cstg-dervied token as well